File: web/src/modes/attract.ts
import type { Rng } from '../core/rng';
import type { Run } from '../game/run';
import type { Track } from '../game/track';

/**
 * Menu attract mode: an endless rolling ride generated ahead of the rider. Terrain is a drifting
 * sum of sines whose downhill bias adapts to the rider's speed, with the odd kicker for airtime.
 * Crashes are part of the show; the game restarts the demo with the next environment.
 */
export class AttractDirector {
  private frontier: number;
  private lastY: number;
  private drift = 0.1;
  private readonly p1: number;
  private readonly p2: number;
  private lastPrune = 0;

  constructor(
    private readonly track: Track,
    private readonly rng: Rng,
  ) {
    this.p1 = rng.range(0, Math.PI * 2);
    this.p2 = rng.range(0, Math.PI * 2);
    track.start = { x: 0, y: 0 };
    track.addLine({ x1: -80, y1: 6, x2: 60, y2: 14, layer: 'level' });
    this.frontier = 60;
    this.lastY = 14;
    for (let i = 0; i < 8; i++) this.chunk(6);
  }

  private chunk(speed: number): void {
    const seg = 40;
    const target = speed > 9.5 ? -0.05 : speed < 4 ? 0.22 : 0.1;
    this.drift += (target - this.drift) * 0.5;
    const kicker = speed >= 5 && speed <= 9 && this.rng.chance(0.2);
    let x = this.frontier;
    let y = this.lastY;
    for (let i = 0; i < 8; i++) {
      const nx = x + seg;
      const slope = this.drift + 0.17 * Math.cos(nx / 260 + this.p1) + 0.16 * Math.cos(nx / 95 + this.p2);
      const ny = y + slope * seg;
      const material = slope > 0.22 && this.rng.chance(0.2) ? 'accel' : 'normal';
      this.track.addLine({ x1: x, y1: y, x2: nx, y2: ny, layer: 'level', material });
      x = nx;
      y = ny;
    }
    if (kicker) {
      const rx = x + 36;
      const ry = y - 12;
      this.track.addLine({ x1: x, y1: y, x2: rx, y2: ry, layer: 'level' });
      x = rx + 120;
      y = ry + 50;
      this.track.addLine({ x1: x, y1: y, x2: x + 240, y2: y + 62, layer: 'level' });
      x += 240;
      y += 62;
    }
    this.frontier = x;
    this.lastY = y;
  }

  /** Extend terrain ahead of the rider and drop lines far behind. */
  update(run: Run): void {
    const c = run.rider.center();
    const v = run.rider.velocity();
    const speed = Math.sqrt(v.x * v.x + v.y * v.y);
    while (this.frontier < c.x + 1600) this.chunk(speed);
    if (run.frame - this.lastPrune >= 40) {
      this.lastPrune = run.frame;
      for (const l of [...this.track.lines.values()]) {
        if (Math.max(l.x1, l.x2) < c.x - 2500) this.track.removeLine(l.id);
      }
    }
  }
}

File: Assets/Scripts/Core/Game/Attract.cs
using System;
using System.Collections.Generic;

namespace CyberRider.Core
{
    /// <summary>
    /// Menu attract mode: an endless rolling ride generated ahead of the rider. Terrain is a drifting
    /// sum of cosines whose downhill bias adapts to the rider's speed, with the odd kicker for airtime.
    /// Crashes are part of the show; the game restarts the demo with the next environment.
    /// </summary>
    public sealed class AttractDirector
    {
        private double _frontier;
        private double _lastY;
        private double _drift = 0.1;
        private readonly double _p1;
        private readonly double _p2;
        private int _lastPrune;
        private readonly Track _track;
        private readonly Rng _rng;

        public AttractDirector(Track track, Rng rng)
        {
            _track = track;
            _rng = rng;
            _p1 = rng.Range(0, Math.PI * 2);
            _p2 = rng.Range(0, Math.PI * 2);
            track.Start = new Vec2d(0, 0);
            track.AddLine(-80, 6, 60, 14, MaterialId.Normal, false, 1, LineLayer.Level);
            _frontier = 60;
            _lastY = 14;
            for (int i = 0; i < 8; i++) Chunk(6);
        }

        private void Chunk(double speed)
        {
            const double seg = 40;
            double target = speed > 9.5 ? -0.05 : speed < 4 ? 0.22 : 0.1;
            _drift += (target - _drift) * 0.5;
            bool kicker = speed >= 5 && speed <= 9 && _rng.Chance(0.2);
            double x = _frontier;
            double y = _lastY;
            for (int i = 0; i < 8; i++)
            {
                double nx = x + seg;
                double slope = _drift + 0.17 * Math.Cos(nx / 260 + _p1) + 0.16 * Math.Cos(nx / 95 + _p2);
                double ny = y + slope * seg;
                MaterialId material = slope > 0.22 && _rng.Chance(0.2) ? MaterialId.Accel : MaterialId.Normal;
                _track.AddLine(x, y, nx, ny, material, false, 1, LineLayer.Level);
                x = nx;
                y = ny;
            }
            if (kicker)
            {
                double rx = x + 36;
                double ry = y - 12;
                _track.AddLine(x, y, rx, ry, MaterialId.Normal, false, 1, LineLayer.Level);
                x = rx + 120;
                y = ry + 50;
                _track.AddLine(x, y, x + 240, y + 62, MaterialId.Normal, false, 1, LineLayer.Level);
                x += 240;
                y += 62;
            }
            _frontier = x;
            _lastY = y;
        }

        /// <summary>Extend terrain ahead of the rider and drop lines far behind.</summary>
        public void Update(Run run)
        {
            Vec2d c = run.Rider.Center();
            Vec2d v = run.Rider.Velocity();
            double speed = Math.Sqrt(v.X * v.X + v.Y * v.Y);
            while (_frontier < c.X + 1600) Chunk(speed);
            if (run.Frame - _lastPrune >= 40)
            {
                _lastPrune = run.Frame;
                foreach (LineData l in new List<LineData>(_track.Lines.Values))
                {
                    if (Math.Max(l.X1, l.X2) < c.X - 2500) _track.RemoveLine(l.Id);
                }
            }
        }
    }
}
